packages/wallet/src/rollup-ovm.ts

```diff
@@ -0,0 +1,116 @@
+/* External Imports */
+import {
+  AndDecider,
+  BigNumber,
+  Decider,
+  Decision,
+  DefaultSignatureVerifier,
+  getLogger,
+  ImplicationProofItem,
+  MerkleInclusionProofDecider,
+  SignatureVerifier,
+  SignedByDBInterface,
+} from '@pigi/core'
+
+/* Internal Imports */
+import {
+  abiEncodeState,
+  abiEncodeStateReceipt,
+  Address,
+  SignedStateReceipt,
+  StateReceipt,
+} from './index'
+
+const log = getLogger('rollup-ovm')
+
+/**
+ * The OVM processor for UniPig-like applications
+ */
+export class RollupOVM {
+  constructor(
+    private readonly signedByDB: SignedByDBInterface,
+    private readonly signedByDecider: Decider,
+    private readonly merkleInclusionDecider: Decider = new MerkleInclusionProofDecider(),
+    private readonly signatureVerifier: SignatureVerifier = DefaultSignatureVerifier.instance()
+  ) {}
+
+  /**
+   * Stores the SignedStateReceipt
+   * @param signedReceipt The signed receipt
+   */
+  public async storeSignedStateReceipt(
+    signedReceipt: SignedStateReceipt
+  ): Promise<void> {
+    const signer = this.signatureVerifier.verifyMessage(
+      abiEncodeStateReceipt(signedReceipt.stateReceipt),
+      signedReceipt.signature
+    )
+    await this.signedByDB.storeSignedMessage(
+      Buffer.from(signedReceipt.signature),
+      Buffer.from(signer)
+    )
+  }
+
+  /**
+   * Determines whether or not the provided StateReceipt is valid, checking that
+   * there is a signature for it, and it has a valid inclusion proof.
+   * @param stateReceipt
+   * @param signer
+   */
+  public async isStateReceiptProvablyValid(
+    stateReceipt: StateReceipt,
+    signer: Address
+  ): Promise<boolean> {
+    // TODO: should we return the decision here and make the caller handle disputes,
+    //       or should this handle disputes?
+    //       I'm leaning toward this handling disputes.
```

Comment:
Yes I like this handling disputes as well

Reply:
Agreed. for now there's a public `getFraudProof(...)` function for getting a fraud  proof. When we hook this bad boy up to mainnet, we'll hook into that to get proof for a dispute and make it private if not used externally.